designs/provisioner-priority.md

```diff
@@ -0,0 +1,268 @@
+# Prioritizing Provisioners
+
+## Goals
+
+- Allowing a user to describe a logical ordering to their provisioners
+- Allowing a user to set defaults and/or fallback orderings for their provisioners
+
+## Background
+
+Currently, there is no logical ordering to how the Karpenter scheduling receives and processes Provisioners. Provisioners are processed in the order they are received from the Kubernetes LIST API, which is randomly. Like pod affinities, users want to convey a preferential relationship to some Provisioners over others. Rather than having to manually map a pod to a specific Provisioner through a `nodeSelector` or `affinity` (which could be cumbersome for clusters with large workloads), a user should be able to specify global preferences to pod assignment to provisioners.
+
+## Use Cases
+
+1. Creating a strongly preferred default provisioner that will always be attempted first, except when more specific configuration is required. This enables scenarios like specifying `kubernetes.io/arch=arm64` for workloads that require this architecture, while preferring `kubernetes.io/arch=amd64` on all other workloads, even if the pod does not contain specific `nodeSelector` constraints. This is particularly useful for those that have high numbers of workloads that only work on `amd64` instances, but they do not want to have to go and assign the specific architecture-specific `nodeSelector` to each of these workloads.
+
+   **Example**
+
+    ```yaml
+    # Default provisioner that will be attempted to be scheduled first
+
+    apiVersion: karpenter.sh/v1alpha5
+    kind: Provisioner
+    metadata:
+      name: default
+    spec:
+      weight: 100
+      requirements:
+      - key: kubernetes.io/arch
+        operator: In
+        values: ["amd64"]
+    ```
+
+    ```yaml
+    # ARM-64 specific provisioner that will be scheduled if the other provisioner does not fit the constraints for this provisioner
+
+    apiVersion: karpenter.sh/v1alpha5
+    kind: Provisioner
+    metadata:
+      name: arm64
+    spec:
+      weight: 50
+      requirements:
+      - key: kubernetes.io/arch
+        operator: In
+        values: ["arm64"]
+    ```
+
+2. Creating a strongly preferred default provisioner for enterprise users that have `reserved` instance types and want to prefer having these instance types scheduled prior to other instance types.
+
+   **Example**
+
+    ```yaml
+    # I have c5.large instance types as reserved instances, so I want to schedule to these nodes first
```

Comment:
Does this make sense?  Don't you have a fixed number of RI and this will keep preferring that regardless until they fail to launch.

Reply:
At the point that it fails to launch this instance type due to ICE, we will null it out in our `getInstaceTypes` call so at that point it should no longer recommend this instance, right? What is the issue with this being the user experience?

Reply:
The experience isn't great, but there's not a great way around it with RIs. You wouldn't get ICE'd once you run out, you'd just lose your discount.

Reply:
Magic: Put a CPU limit on the provisioner that only lets you launch up to the # of RIs you have